src/enzyme_ad/jax/Implementations/HLODerivatives.td

```diff
@@ -1446,3 +1444,158 @@ def : HLODerivative<"FftOp",
+    mapping[to_i64(dim)] = i;
+  }
+
+  int next = op.getBroadcastDimensions().size();
+  for (int i = 0; i < outTy.getRank(); i++) {
+    if (mapping[i] == -1) {
+      mapping[i] = next++;
+    }
+  }
+
+  for (int i = 0; i < outTy.getRank(); i++) {
+    perm[mapping[i]] = i;
+  }
+  getI64Attr(builder, perm);
+}]>;
+
+def InsertDeletedReduceDimsType : GlobalExpr</*needsprimal*/0, /*needsshadow*/0, [{
+  SmallVector<int64_t> reduceDims;
+  auto outTy = cast<RankedTensorType>(op.getType());
+  auto bcastDims = op.getBroadcastDimensions();
+  auto inTy = cast<RankedTensorType>(op.getOperand().getType());
+  auto outShape = outTy.getShape();
+
+  for (auto en : llvm::enumerate(outTy.getShape())) {
+    ssize_t bcastIdx = -1;
+    for (auto en2 : llvm::enumerate(bcastDims)) {
+      if (en2.value() == en.index()) {
+        bcastIdx = en2.index();
+        break;
+      }
+    }
+    if (bcastIdx != -1) {
+      if (en.value() != inTy.getShape()[bcastIdx]) {
+        reduceDims.push_back(en.index());
+        assert(inTy.getShape()[bcastIdx] == 1);
+      }
+      continue;
+    }
+    reduceDims.push_back(en.index());
+  }
+
+  SmallVector<int64_t> reshapeShape(outTy.getRank(), -1);
+  for (auto [i, sz] : llvm::enumerate(outShape)) {
+    if (llvm::is_contained(reduceDims, i)) {
+      reshapeShape[i] = 1;
+    } else {
+      reshapeShape[i] = sz;
+    }
+  }
+
+  if (gutils->width > 1) {
+    reshapeShape.insert(reshapeShape.begin(), gutils->width);
+  }
+  RankedTensorType::get(reshapeShape, outTy.getElementType());
+}]>;
+
+def ReduceAdd : HLOInst<"ReduceOp", ")->getResult(0)", "createAddRegion(">;
+
+def ResultTypeWithBatch : GlobalExpr</*needsprimal*/0, /*needsshadow*/0, [{
+  auto outTy = cast<RankedTensorType>(op.getType());
+  auto outShape = outTy.getShape();
+  SmallVector<int64_t> newShape(outShape.begin(), outShape.end());
+  if (gutils->width > 1) {
+    newShape.insert(newShape.begin(), gutils->width);
+  }
+  RankedTensorType::get(newShape, outTy.getElementType());
+}]>;
+
+def InputTypeWithBatch : GlobalExpr</*needsprimal*/0, /*needsshadow*/0, [{
+  auto inTy = cast<RankedTensorType>(op.getOperand().getType());
+  auto inShape = inTy.getShape();
+  SmallVector<int64_t> newShape(inShape.begin(), inShape.end());
+  if (gutils->width > 1) {
+    newShape.insert(newShape.begin(), gutils->width);
+  }
+  RankedTensorType::get(newShape, inTy.getElementType());
+}]>;
+
+def : HLODerivative<"BroadcastInDimOp", (Op $x),
+      [
+        (
+          Reshape
+          (InputTypeWithBatch),
+          (
+            Transpose
+            (
+              Reshape
+              (InsertDeletedReduceDimsType),
+              (ReduceAdd (DiffeRet), (ReduceAddNullValue), (BroadcastDimsToReductionDims))
+            ),
+            (BroadcastDimensionsToInversePermutation)
+          )
+        )
+      ],
+      (
+        BroadcastInDim (ResultTypeWithBatch), (Shadow $x), (getBroadcastDimensionsWithBatch)
```

Comment:
cam you explicitly add batch tests? In principle the tablegen interface should automatically do batching on top of so we should check it doesn't accidentally conflict